src/Data/Text/Foreign.hs

```diff
@@ -149,6 +151,23 @@ asForeignPtr t@(Text _arr _off len) = do
   unsafeWithForeignPtr fp $ unsafeCopyToPtr t
   return (fp, I8 len)
 
+-- | Marshal a 'Text' into a C string with a trailing NUL byte,
+-- encoded as UTF-8 in temporary storage.
+--
+-- The temporary storage is freed when the subcomputation terminates
+-- (either normally or via an exception), so the pointer to the
+-- temporary storage must /not/ be used after this function returns.
+--
+-- @since 2.3
+withCString :: Text -> (CString -> I8 -> IO a) -> IO a
```

Comment:
Any pointers to functions with a similar design in other packages?

Reply:
`I8` doesn't look like a right type for the length.

`bytestring` has https://hackage.haskell.org/package/bytestring-0.11.3.1/docs/Data-ByteString-Char8.html#v:useAsCStringLen, with ` (CStringLen -> IO r) -> IO r` type. That however doesn't add a zero.

`base` has https://hackage.haskell.org/package/base-4.16.1.0/docs/Foreign-C-String.html#v:withCStringLen.
Its `withCString` adds a NUL byte, and `withCStringLen` doesn't. Interesting choice.

Having a variant with length and trailing zero make sense, and the buffer is copied anyway, so it's cheap to just add zero.

Reply:
EDIT: yeah, `text` actually has `withCStringLen` https://hackage.haskell.org/package/text-2.0/docs/Data-Text-Foreign.html#v:withCStringLen which doesn't add a null, and **doesn't** have `withCString` which adds a null and doesn't give you a length. But IMHO it makes sense to have all three variants:

1. `withCString :: Text -> (CString -> IO r) -> IO r` null-terminated
2. `withCStringLen :: Text -> (CStringLen -> IO r) -> IO r` with length, but not null terminated
3. `withCStringLenAndNul :: Text -> (CStringLen -> IO r) -> IO r` with length but also null terminated.

Reply:
IMO using `CStringLen` for null-terminated c-strings is confusing. 

GHC uses this type distinction to mark 0-terminated strings differently from explicitly sized arrays:
https://hackage.haskell.org/package/base-4.16.1.0/docs/Foreign-C-String.html#v:withCString

Reply:
> I8 doesn't look like a right type for the length.

Ugh.. I thought it was made specifically to mark length-in-bytes to prevent confusing it with length-in-whatever-the-user-thinks-it-is. But now I see it has no constructor published, forcing users to jump through the hoops.

Reply:
@dpwiz different type?

```haskell
type CString = Ptr CChar
type CStringLen = (Ptr CChar, Int)
```

They are just aliases to `Ptr CChar`.

Reply:
Different names, whatever. The point being, there is a convention that `CString` is *used* in different places than `CStringLen`. Adding things like `withCStringLen0` disrupts this convention.
According to it `withCStringLenAndNul` should be `withCStringAndLen`.

Reply:
A situation I'm worried about goes like this:

Some imaginary lib uses the following functions:

- `Widgets.useNulTerminatedUtf8 :: CString -> IO ()`
- `Widgets.useSizedUtf8 :: CStringLen -> IO ()`

Now, what would be a correct pick?

- `Text.withCString :: Text -> (CString -> IO ()) -> IO ()` -- obvious, `useNulTerminatedUtf8`
- `Text.withCStringLen :: Text -> (CStringLen -> IO ()) -> IO ()` -- obvious, `useSizedUtf8`
- `Text.withCStringLenAndNul :: Text -> (CStringLen -> IO ()) -> IO ()` -- who to believe here, the function name or the type name? Ignore both and actually read the docs.

Reply:
@dpwiz but this, your PR adds `??? :: Text -> (CString -> Int -> IO r) -> IO r` (length + null) variant, with a name of `Text -> (CString -> IO r) -> IO r` function (if we are to follow naming conventions of other libraries).

Please amend the PR to what you actually want. Currently either the name or the type are wrong.

Reply:
Done